flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/rules/logical/PushPartitionIntoTableSourceScanRule.scala

```diff
@@ -0,0 +1,166 @@
+          case _ => false
+        }
+      case _ => false
+    }
+  }
+
+  override def onMatch(call: RelOptRuleCall): Unit = {
+    val filter: Filter = call.rel(0)
+    val scan: LogicalTableScan = call.rel(1)
+    val table: FlinkRelOptTable = scan.getTable.asInstanceOf[FlinkRelOptTable]
+    pushPartitionIntoScan(call, filter, scan, table)
+  }
+
+  private def pushPartitionIntoScan(
+      call: RelOptRuleCall,
+      filter: Filter,
+      scan: LogicalTableScan,
+      relOptTable: FlinkRelOptTable): Unit = {
+
+    val tableSourceTable = relOptTable.unwrap(classOf[TableSourceTable[_]])
+    val tableSource = tableSourceTable.tableSource.asInstanceOf[PartitionableTableSource]
+    val partitionFieldNames = tableSource.getPartitionFieldNames.toList.toArray
+    val inputFieldType = filter.getInput.getRowType
+
+    val relBuilder = call.builder()
+    val maxCnfNodeCount = FlinkRelOptUtil.getMaxCnfNodeCount(scan)
+    val (partitionPredicate, nonPartitionPredicate) =
+      RexNodeExtractor.extractPartitionPredicates(
+        filter.getCondition,
+        maxCnfNodeCount,
+        inputFieldType.getFieldNames.toList.toArray,
+        relBuilder.getRexBuilder,
+        partitionFieldNames
+      )
+
+    if (partitionPredicate.isAlwaysTrue) {
+      // no partition predicates in filter
+      return
+    }
+
+    val finalPartitionPredicate = adjustPartitionPredicate(
+      inputFieldType.getFieldNames.toList.toArray,
+      partitionFieldNames,
+      partitionPredicate
+    )
+    val partitionFieldTypes = partitionFieldNames.map { name =>
+      val index = inputFieldType.getFieldNames.indexOf(name)
+      require(index >= 0, s"$name is not found in ${inputFieldType.getFieldNames.mkString(", ")}")
+      inputFieldType.getFieldList.get(index).getType
+    }.map(FlinkTypeFactory.toLogicalType)
+
+    val allPartitions = tableSource.getPartitions
+    val remainingPartitions = PartitionPruner.prunePartitions(
+      call.getPlanner.getContext.asInstanceOf[FlinkContext].getTableConfig,
+      partitionFieldNames,
+      partitionFieldTypes,
+      allPartitions,
+      finalPartitionPredicate
+    )
+
+    val newTableSource = tableSource.applyPartitionPruning(remainingPartitions)
+
+    val statistic = tableSourceTable.statistic
+    val newStatistic = if (remainingPartitions.size() == allPartitions.size()) {
+      // Keep all Statistics if no predicates can be pushed down
+      statistic
+    } else if (statistic == FlinkStatistic.UNKNOWN) {
+      statistic
+    } else {
+      // Remove tableStats after predicates pushed down
+      FlinkStatistic.builder().statistic(statistic).tableStats(null).build()
+    }
+    val newTableSourceTable = new TableSourceTable(
+      newTableSource, tableSourceTable.isStreaming, newStatistic)
+    val typeFactory = filter.getCluster.getTypeFactory
+   val newRelOptTable = relOptTable.copy(
+     newTableSourceTable, tableSourceTable.getRowType(typeFactory))
```

Comment:
val newRelOptTable = relOptTable.copy(newTableSourceTable, relOptTable.getRowType))?
since type of RelOptTable will not change after partition push down.

Reply:
inputFieldType could also be used here.